doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+
+## Use cases
+
+### Use with password-only authentication
+
+The user can set multiple passwords per account. Each password can be then used
+for authenticating on only one device. Various passwords can as well be used to
+authenticate to various services or their sets. These divisions can even be
+combined, thus - in the most extreme case - having one password for
+authenticating to one specific service on one particular device. Granularity of
+such divisions will be fully up to user's discretion.
+
+### Use as a part of 2FA (OTP)
+
+This may work, as the code handling the second factor is independent from the
+code handling the password(s), but it will not be a goal for this extension nor
+will it be tested.
+
+~~The user could enter an application-specific password instead of the primary
+one, and then an OTP, allowing application-specific passwords to be used as
+a part of 2FA. Although 2FA reduces the risk of account hijacking, the benefits
+of application-specific passwords still improve security in this case as well.~~
+
+### Use as a fallback mechanism when 2FA is activated, but the client
+    application does not support it
+
+This is almost certainly not needed, as authentication is managed by FreeIPA,
+which obviously supports 2FA, and client applications only consume the result of
+the authentication. And again, 2FA is out of scope of this extension.
+
+### Use as part of RADIUS and Hardened Password authentication mechanisms
+
+These may work, as the code handling these authentication mechanisms is
+independent from the code handling the password(s), but it will not be a goal
+for this extension nor will it be tested.
+
+~~As both of these authentication mechanisms include sending a password for
+authentication, they could also be used with application-specific passwords.
+As pure RADIUS provides only weak protection of the sent password, it might
+be even more desirable than in the case of traditional password
+authentication.~~
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+Application-specific passwords will be stored in a new LDAP subtree in the form
+```
+object,cn=appname,cn=apps,$SUFFIX.
+```
+Every application will have its own subtree in which users will be able to
+create their objects if they will want to use that application - this means
+objects containing application-specific passwords will not be grouped by users,
+but by applications. There will be one object for each application-specific
+password of each user; one user will be able to have one password for a given
+application. The object containing an application-specific password will have
+the following attributes:
+* `krbprincipalkey` - will hold the application-specific password itself
+* `krbprincipalname` - two-component Kerberos principal (see below)
+* backlink to the owning user (that created it) - this backlink must be
+  maintained; inspiration can be seen in working with hostgroups
```

Comment:
We also need to specify clearly what is allowed and what is not, how password policies will apply (or not, hint they probably shouldn't by default), we need to decide if userPassword needs to be present (almost certainly it will need to be for applications that want to bind via LDAP).

Reply:
@tiran As I was further thinking about it, I expected it to be `memberof` attribute that would contain the username of the owning user. Its maintanance would involve only changing it in accordance with the change of the owning user's username. It would also be used for finding the object in the subtree during checking whether the password the user entered matches. So is this scenario suboptimal? As I have now quickly looked on `ipatokenowner` attribute of the `otptoken` class as well as work with it, I found how it is automatically filled when not specified by the user, as well as how the `managedby` attribute is modified thogether with it - however, I have not found what is being done with it when the owning user's username changes - it seems that this is not addressed in the baseuser and user plugins.

@simo5 Whatever exact form this attribute will have, I expect it to be managed fully automatically, without any possibility of intervention by the user. As for password policies, this approach makes sense if we expect to always generate app-specific passwords. Then there will be no need for password policies for them at all. As for the `userPassword` attribute, OK, it will be present, and just as as the `krbprincipalkey` attribute, it will function analogically as the attribute of the same name in the `baseuser` class.